File: src/docforge/openapi/loader.py
import json
from pathlib import Path
from typing import Any

import yaml


class UnsupportedSpecError(ValueError):
    """Raised when the spec file is not a supported OpenAPI 3.x document."""


def load_spec(path: Path) -> dict[str, Any]:
    suffix = path.suffix.lower()
    raw = path.read_text(encoding="utf-8")

    if suffix == ".json":
        spec = json.loads(raw)
    elif suffix in (".yaml", ".yml"):
        spec = yaml.safe_load(raw)
    else:
        raise UnsupportedSpecError(
            f"unknown spec suffix {suffix!r} (expected .json/.yaml/.yml)"
        )

    if not isinstance(spec, dict):
        raise UnsupportedSpecError("spec root must be an object")

    if "swagger" in spec:
        raise UnsupportedSpecError(
            f"Swagger 2.0 not supported (found swagger={spec['swagger']!r}); "
            "convert to OpenAPI 3.x first"
        )

    version = spec.get("openapi", "")
    if not isinstance(version, str) or not version.startswith("3."):
        raise UnsupportedSpecError(
            f"unsupported openapi version: {version!r} (expected 3.x)"
        )

    return spec

File: src/openapi/loader.ts
import { readFileSync } from "node:fs";
import { extname } from "node:path";
import { load as yamlLoad } from "js-yaml";

import { fetchUrl, type FetchOptions } from "../http/fetch.js";

export class UnsupportedSpecError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnsupportedSpecError";
  }
}

export function loadSpec(path: string): Record<string, unknown> {
  const suffix = extname(path).toLowerCase();
  const raw = readFileSync(path, "utf8");

  let spec: unknown;
  if (suffix === ".json") {
    spec = JSON.parse(raw);
  } else if (suffix === ".yaml" || suffix === ".yml") {
    spec = yamlLoad(raw);
  } else {
    throw new UnsupportedSpecError(
      `unknown spec suffix '${suffix}' (expected .json/.yaml/.yml)`,
    );
  }

  if (
    spec === null ||
    typeof spec !== "object" ||
    Array.isArray(spec)
  ) {
    throw new UnsupportedSpecError("spec root must be an object");
  }

  const obj = spec as Record<string, unknown>;

  if ("swagger" in obj) {
    throw new UnsupportedSpecError(
      `Swagger 2.0 not supported (found swagger=${JSON.stringify(obj.swagger)}); ` +
        "convert to OpenAPI 3.x first",
    );
  }

  const version = obj.openapi;
  if (typeof version !== "string" || !version.startsWith("3.")) {
    throw new UnsupportedSpecError(
      `unsupported openapi version: ${JSON.stringify(version)} (expected 3.x)`,
    );
  }

  return obj;
}

export async function loadSpecFromUrl(
  url: string,
  opts: FetchOptions,
): Promise<Record<string, unknown>> {
  const res = await fetchUrl(url, opts);
  const ct = res.contentType.toLowerCase();
  const body = res.bytes.toString("utf8");
  let spec: unknown;
  if (ct.includes("json")) {
    spec = JSON.parse(body);
  } else {
    spec = yamlLoad(body);
  }
  if (spec === null || typeof spec !== "object" || Array.isArray(spec)) {
    throw new UnsupportedSpecError("spec root must be an object");
  }
  const obj = spec as Record<string, unknown>;
  if ("swagger" in obj) {
    throw new UnsupportedSpecError(
      `Swagger 2.0 not supported (found swagger=${JSON.stringify(obj.swagger)}); convert to OpenAPI 3.x first`,
    );
  }
  const version = obj.openapi;
  if (typeof version !== "string" || !version.startsWith("3.")) {
    throw new UnsupportedSpecError(
      `unsupported openapi version: ${JSON.stringify(version)} (expected 3.x)`,
    );
  }
  return obj;
}
